xprof/tally.hpp

```diff
@@ -1,31 +1,820 @@
+      return s.substr(18, s.size());
+    return s;
+  }
+  return mangle_name;
+}
+
+//! Returns a number as a string with the given number of decimals.
+//! @param a_value the value to be casted to string with the given decimal places.
+//! @param units units to be append at the end of the string.
+//! @param n number of decimal places required.
+//! REFERENCE: https://stackoverflow.com/questions/16605967/set-precision-of-stdto-string-when-converting-floating-point-values
+template <typename T>
+std::string to_string_with_precision(const T a_value, const std::string units, const int n = 2) {
+  std::ostringstream out;
+  out.precision(n);
+  out << std::fixed << a_value << units;
+  return out.str();
+}
+
+//! TallyCoreBase is a callbacks duration data collection and aggregation helper.
+//! It is of interest to collect data for every (host,pid,tid,api_call_name) entity.
+//! Since the same entity can take place several times, i.e., a thread spawned from 
+//! a process running in a given host can call api_call_name several times, our 
+//! interest is to aggregate these durations in a single one per entity.
+//! In addition to the duration, othe data of interes is collected from different
+//! ocurrences of the same entity such as, what was the minumum and max durations
+//! among the ocurrences, the number of times an api_call_name happed for a given 
+// "htp" (host,pid,tid), and how many occurrences failed.
+//! Once the data of an entity is collected, when considering all its ocurrences,
+//! This helper calls facilitates aggregation of data by overloading += and + operators.
+class TallyCoreBase {
+public:
+  TallyCoreBase() {}
+
+  TallyCoreBase(uint64_t _dur, uint64_t _err) : duration{_dur}, error{_err} {
+    count = 1;
+    if (!error) {
+      min = duration;
+      max = duration;
+    } else
+      duration = 0;
+  }
+
+  //! Total exeuction time 
+  //! The sum of the durations of traces of an specific entity (host,proces,thread,api_call_name), 
+  // it is ("Time" in statistics)
+  uint64_t duration{0};
+  
+  //! Accumulates the amount of errors found for an (host,pid,tid,api_call_name) entity.
+  uint64_t error{0};
+
+  //! Minimum duration found among occurrences of the same (host,pid,tid,api_call_name) entity.
+  uint64_t min{std::numeric_limits<uint64_t>::max()};
+
+  //! Maximum duration found among occurrences of the same (host,pid,tid,api_call_name) entity.
+  uint64_t max{0};
+
+  //! Count the number of times an error is found for a given (host,pid,tid,api_call_name) entity.
+  uint64_t count{0};
+
+  //! Percentage of the total execution time.
+  //! duration / app_total_execution_time, this is computed at the end, once we have collected 
+  //! the duratiin information of all the ocurrences of a given (host,pid,tid,api_call_name) entity.
+  double duration_ratio{1.};
+
+  //! Average duration.
+  //! It is rougly duration / #of_sucessfull_calls, see "finalize" member function.
```

Comment:
Do you think we need those comment? They seem pretty obvious to me

Reply:
We actually don't need them, they helped my to understand, now we can delete them.

Reply:
We can keep the other comment; I think they help us understand the complicated C++ code. But those one I think we can delete